### PR TITLE
Tailored Audience for v4

### PR DESCRIPTION
**Pull request by esakai**
This is an implementation for Twitter Ads API v4 endpoints for tailored audiences.

- POST accounts/:account_id/tailored_audiences
- POST accounts/:account_id/tailored_audiences/:tailored_audience_id/users

Tests are not available yet.

### CODE REVIEW

**Comment by esakai** [comment]
@jbabichjapan 
It would be great if you could review this pull request. Thank you in advance.

**Comment by jbabichjapan** [comment]
Thanks so much for this - @tushdante will be reviewing this and also has been preparing the rest of the support for v4 - I think it would be good to align the checkin so that the other functionality doesn't break when we upgrade the API_VERSION from 3 to 4

File: lib/twitter-ads/audiences/tailored_audience.rb [hunk 1]
```diff
@@ -86,6 +88,25 @@ def create(account, file_path, name, list_type)
...
+      # @example
+      #   audience = TailoredAudience.create_instance(account, 'my list')
+      #
+      # @param account [Account] The account object instance.
+      # @param name [String] The tailored audience name.
+      #
+      # @since 4.0
+      #
+      # @return [TailoredAudience] The newly created tailored audience instance.
+      def create_instance(account, name)
```
Comment (tushdante) [inline lib/twitter-ads/audiences/tailored_audience.rb]: Given that v4 will no longer support TON based audiences, I'd suggest removing this method in favor of editing the existing `create` method instead

File: lib/twitter-ads/audiences/tailored_audience.rb [hunk 2]
```diff
@@ -194,6 +215,49 @@ def self.memberships(account, params)
...
+    #           ]
+    #         }
+    #       }
+    #     ]
+    #   )
+    #
+    # @since 4.0
+    #
+    # @return success_count, total_count
+    def users(params)
```
Comment (tushdante) [inline lib/twitter-ads/audiences/tailored_audience.rb]: Please add `@param params [JSON object] A hash containing the list of users to be added/removed/updated` as a comment to describe the input format

File: lib/twitter-ads/client.rb [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 module TwitterAds
 
-  API_VERSION = '3'.freeze
+  API_VERSION = '4'.freeze
```
Comment (tushdante) [inline lib/twitter-ads/client.rb]: Lets leave this as `3` for now, given that this change will break all deprecated endpoints. We'll make the change once all the other v4 changes have been merged.

**Comment by tushdante** [comment]
Left a couple comments @esakai ! Let me know if there's any additional questions. The changes should be simple enough and once we've merged this PR we can make the change to v4.

File: lib/twitter-ads/audiences/tailored_audience.rb (hunk 1 above)
Comment (esakai) [reply lib/twitter-ads/audiences/tailored_audience.rb]: @tushdante 

There is a concern. The signatures of these two methods are totally different.  

```
create(account, file_path, name, list_type)
create_instance(account, name)
```

If I made a method like `create(account, name)`, the backward compatibility would be broken. Is this OK with you? If so, I would simply drop the current `create` method and rename `create_instance` to `create`.

File: lib/twitter-ads/audiences/tailored_audience.rb (hunk 2 above)
Comment (esakai) [reply lib/twitter-ads/audiences/tailored_audience.rb]: @tushdante 
Thank you for your comment. I will add it.

File: lib/twitter-ads/audiences/tailored_audience.rb (hunk 1 above)
Comment (tushdante) [reply lib/twitter-ads/audiences/tailored_audience.rb]: Good point! I think it should be fine to just drop the current `create` method and rename the `create_instance` to `create`, given that the current `create` method will no longer be supported as of v4.

Comment (esakai) [reply lib/twitter-ads/audiences/tailored_audience.rb]: @tushdante Thank you for your comment. I understand. I will modify the code as you mentioned.

**Comment by eccyan** [comment]
How's this PR going?

**Comment by esakai** [comment]
I have already implemented all parts I was requested to change. I am waiting for @tushdante 's review.

**Comment by treby** [comment]
It seems the code conflicts with master.
@esakai do you plan to resolve that?

**Comment by esakai** [comment]
@treby 

Thank you for your notice.
I have just rebased this branch on the master branch so that the conflict can be resolved.

There is one more suggestion.
I think that `TwitterAds::TailoredAudience.memberships` is no longer needed.
Do you want me to delete the method?
If you wish to do so, you can delete it yourself.
Please let me know your thoughts.

**Comment by treby** [comment]
> I think that `TwitterAds::TailoredAudience.memberships` is no longer needed.

I think so too.
According to twitter, version 3 of the Ads API reached end of life on February 28, 2019.

The document for the API is no longer available, too.
https://developer.twitter.com/en/docs/ads/audiences/api-reference/real-time

I can't decide the scope of this PR (Maybe @jbabichjapan or @tushdante - gem maintainers can), but just in my opinion both of the way (below) are ok.

- option 1: First merge this PR, remove outdated feature on other PR
- option 2: Including outdated feature removal on this PR, and review & merge

To maintainers, what do you think?

**Comment by smaeda-ks** [comment]
@treby @esakai Thanks. As you noticed, some of the old Audience endpoints are no longer work in v5 and there are even more lines to be removed. We will take care that part in another PR, so let's freeze this PR for now and merge it as it's been a while since created last year. Though, thank you for your patience :)

cc: @tushdante

**Comment by treby** [comment]
@esakai I think this PR is ready to merge (Perhaps you need to follow master branch :) ).
Do you have any works for this PR before merging it?

**Comment by esakai** [comment]
@treby 
Thank you for your reminder.
I have just rebased this branch on `twitterdev:master` branch.
I have already done everything.
You can merge this branch to the master branch now.

**Comment by esakai** [comment]
I think that this pull request should be merged because all the problems are solved now. I am wondering why this pull request is still open.

**Comment by bugcloud** [comment]
What is the `blocker` of this PR??

**Comment by smaeda-ks** [comment]
@esakai @treby @bugcloud This is now merged. Again, thanks for your contribution!

**Pull request merged**